### DISCUSSION TITLE
Is it possible to re-generate the `vendor/craftcms/plugins.php` file?

### DISCUSSION THREAD

**Discussion by mmikkel** [craftcms/cms#12064] [Q&A] [answered]
On a few different sites, I've got some plugins that are long-since uninstalled, but that still appear under Settings -> Plugins in the CP and in the list generated by `php craft plugin/list`.

These plugins were both Composer-uninstalled and Craft-uninstalled at some point, but for some reason Craft never removed their references in the `vendor/craftcms/plugins.php` file, which is the reason why they still show up.

For reference, here's how these plugins appear (note the missing icon since their Composer packages are no longer in `vendor`):

![CleanShot 2022-10-05 at 13 03 28](https://user-images.githubusercontent.com/298510/194046780-c24c34c6-2f53-49bd-b750-e83f0313f5ba.png)

As these are very old sites and I haven't seen the issue crop up on newer installs, I don't really care how this happened. But, I _would_ like to update the `plugins.php` file somehow, removing any references to the non-existent plugins.

I've tried running `php craft plugin/uninstall {handle} --force` and also hitting the "Remove" button in the CP, but neither method updated the `plugins.php` file. I also tried deleting the `plugins.php` file in the hope that Craft would re-generate it on pageload; it doesn't and the site dies.

_I assume re-installing and re-uninstalling the relevant packages would do the trick, but since there are several dependency conflicts with currently installed plugins, this is not a feasible solution either._

Apart from manually editing the `plugins.php` file or nuking the `vendor` folder, is there any way to make Craft re-generate it from scratch and/or purge references to the uninstalled plugins? (I tried to figure out how and when the plugin installer _initially_ creates that file, but I couldn't spot it in the source 🤷)

**Comment by brandonkelly** [comment, marked as answer]
That file is generated by the [`craftcms/plugin-installer`](https://github.com/craftcms/plugin-installer) Composer plugin, which registers a new installer type for `craft-plugin` packages.

There’s no way to get it to run outside of `composer update`/`install`/`remove`.
